### PR TITLE
Make ICL configurations more concise

### PR DESCRIPTION
**Pull request by abhi-mosaic**
Now you can specify like:

```yaml
icl_tasks:
- label: piqa
  dataset_uri: ./icl_eval/piqa.jsonl
  num_fewshot: [0, 1, 5]
  icl_task_type: multiple_choice
```

and it will default the rest based on the `icl_task_type`. You can still override everything.

@bmosaicml just FYI as I know you are working on #9 .

### CODE REVIEW

File: llmfoundry/data/denoising.py [hunk 1]
```diff
@@ -476,7 +476,7 @@ def build_text_denoising_dataloader(
         download_timeout=cfg.dataset.get('download_timeout', 60),
         validate_hash=cfg.dataset.get('validate_hash'),
         shuffle_seed=cfg.dataset.get('shuffle_seed'),
-        num_canonical_nodes=cfg.dataset.get('num_canonical_nodes'),
+        num_canonical_nodes=cfg.dataset.get('num_canonical_nodes', 128),
```
Comment (vchiley) [inline llmfoundry/data/denoising.py:479]: y not just allows this to be `None`?
`None` is the [default in streaming](https://docs.mosaicml.com/projects/streaming/en/stable/api_reference/generated/streaming.StreamingDataset.html)

File: llmfoundry/main.py [hunk 2]
```diff
@@ -101,7 +101,7 @@ def main(cfg):
         f'torch.distributed.*_base is a private function and will be deprecated.*'
     )
 
-    cfg.dist_timeout = cfg.get('dist_timeout', 1800.0)
+    cfg.dist_timeout = cfg.get('dist_timeout', 600.0)
```
Comment (vchiley) [inline llmfoundry/main.py:104]: nit: why is this a float?

File: llmfoundry/common/builders.py [hunk 3]
```diff
@@ -163,23 +163,38 @@ def build_dataloader(cfg, tokenizer, device_batch_size):
...
-        assert 'batch_size' in icl_cfg
-        assert 'metric_names' in icl_cfg
-        assert 'prompt_string' in icl_cfg
-        assert 'example_delimiter' in icl_cfg
-        assert 'continuation_delimiter' in icl_cfg
-        assert 'label' in icl_cfg
+
+        if 'metric_names' not in icl_cfg:
+            if icl_cfg.icl_task_type == 'language_modeling':
+                icl_cfg.metric_names = 'InContextLearningLMAccuracy'
```
Comment (vchiley) [inline llmfoundry/common/builders.py:178]: Should `metric_names` be a list?

**Review by vchiley — COMMENTED** [review]
we could updt the icl eval yamls [here](https://github.com/mosaicml/llm-foundry/tree/main/llmfoundry/icl_eval/yamls)

File: llmfoundry/common/builders.py [hunk 4]
```diff
@@ -163,23 +163,38 @@ def build_dataloader(cfg, tokenizer, device_batch_size):
...
+        if 'example_delimiter' not in icl_cfg:
+            icl_cfg.example_delimiter = '\n'
+        if 'continuation_delimiter' not in icl_cfg:
+            icl_cfg.continuation_delimiter = ' '
+        if 'batch_size' not in icl_cfg:
+            icl_cfg.batch_size = default_batch_size
 
     for icl_cfg in cfg.icl_tasks:
         _validate_cfg(icl_cfg)
+        print(icl_cfg)
```
Comment (dakinggg) [inline llmfoundry/common/builders.py:197]: remove

Comment (vchiley) [reply llmfoundry/common/builders.py:197]: or hide under `if verbose`?

File: llmfoundry/data/denoising.py (hunk 1 above)
Comment (abhi-mosaic) [reply llmfoundry/data/denoising.py:479]: Not safe for elastic determinism, there is an ongoing discussion in Slack but I really want this repo to be deterministic by default.

File: llmfoundry/main.py (hunk 2 above)
Comment (abhi-mosaic) [reply llmfoundry/main.py:104]: I think Composer `dist.initialize_dist` takes a float.

**Review by vchiley — APPROVED** [review]
lgtm

**Pull request merged**